apps/prairielearn/src/lib/submissions-upload.ts

```diff
@@ -0,0 +1,221 @@
+ */
+function makeDedupedInserter<T>() {
+  return memoize(async (key: string[], fn: () => Promise<T>) => await fn(), {
+    cacheKey: (args) => args[0].join(':'),
+  });
+}
+
+/**
+ * Processes a CSV file containing submissions and creates users, assessment instances,
+ * instance questions, variants, and submissions in the database.
+ *
+ * Note that this will delete all existing assessment instances for the assessment.
+ * Use this function with caution! Specifically, it should really only ever be used
+ * in dev mode.
+ */
+export async function uploadSubmissions(
+  assessment_id: string,
+  csvFile: Express.Multer.File | null | undefined,
+  user_id: string,
+  authn_user_id: string,
+): Promise<string> {
+  if (csvFile == null) {
+    throw new Error('No CSV file uploaded');
+  }
+
+  const { assessment_label, course_instance_id, course_id } =
+    await selectAssessmentInfoForJob(assessment_id);
+
+  const serverJob = await createServerJob({
+    courseId: course_id,
+    courseInstanceId: course_instance_id,
+    assessmentId: assessment_id,
+    userId: user_id,
+    authnUserId: authn_user_id,
+    type: 'upload_submissions',
+    description: 'Upload submissions CSV for ' + assessment_label,
+  });
+
+  const selectUser = memoize(async (uid: string) => await selectOrInsertUserByUid(uid));
+  const selectQuestion = memoize(
+    async (qid: string) => await selectQuestionByQid({ course_id, qid }),
+  );
+  const selectAssessmentQuestion = memoize(
+    async (question_id: string) =>
+      await sqldb.queryRow(
+        sql.select_assessment_question,
+        { assessment_id, question_id },
+        AssessmentQuestionSchema,
+      ),
+  );
+
+  serverJob.executeInBackground(async (job) => {
+    job.info('Deleting all existing assessment instances');
+    await deleteAllAssessmentInstancesForAssessment(assessment_id, authn_user_id);
+
+    job.info('Uploading submissions CSV for ' + assessment_label);
+
+    let successCount = 0;
+    let errorCount = 0;
+
+    const getOrInsertAssessmentInstance = makeDedupedInserter<string>();
+    const getOrInsertInstanceQuestion = makeDedupedInserter<string>();
+    const getOrInsertVariant = makeDedupedInserter<string>();
+
+    job.info(`Parsing uploaded CSV file "${csvFile.originalname}" (${csvFile.size} bytes)`);
+    const csvStream = streamifier.createReadStream(csvFile.buffer, {
+      encoding: 'utf8',
+    });
+    const csvConverter = csvtojson();
+    await csvConverter.fromStream(csvStream).subscribe(async (rawJson, number) => {
+      const lineNumber = number + 2;
+      job.verbose(`Processing CSV line ${lineNumber}: ${JSON.stringify(rawJson)}`);
+
+      try {
+        const row = SubmissionCsvRowSchema.parse(rawJson);
+
+        // For simplicity, we're not handling group work until it's needed.
+        if (row['Group name']) throw new Error('Group work is not supported yet');
+
+        const user = await selectUser(row.UID);
```

Comment:
This operation does not enrol the user to the course instance, is that a problem? I don't think it is, just raising it to make sure this is considered.

Reply:
Also, a bit of a nit, but we have the UIN and name from the file, is it worth setting those to the user record? Again, I understand if you believe this is not the focus here.

Reply:
Sounds good. Minor thing about missing enrollments is that the student will be listed with role "None", but this does not actually affect functionality in most cases.

Reply:
Yep. Again, I'm pretty narrowly focused on the immediate needs of local development for things like AI grading. If you or others require changes to this in the future, please request them.